packages/plugin-pnpm/sources/PnpmLinker.ts

```diff
@@ -211,9 +218,37 @@ class PnpmInstaller implements Installer {
     if (this.opts.project.configuration.get(`nodeLinker`) !== `pnpm`) {
       await xfs.removePromise(storeLocation);
     } else {
+      const removals: Array<Promise<void>> = [];
+
       const expectedEntries = new Set<Filename>();
-      for (const packageLocation of this.packageLocations.values())
-        expectedEntries.add(ppath.basename(packageLocation));
+      for (const packageLocation of this.customData.pathByLocator.values()) {
+        const subpath = ppath.contains(storeLocation, packageLocation);
+        if (subpath !== null) {
+          const [storeEntry, /* Filename.nodeModules */, ...identComponents] = subpath.split(ppath.sep);
+          expectedEntries.add(storeEntry as Filename);
+
+          const storeEntryPath = ppath.join(storeLocation, storeEntry as Filename);
+
+          removals.push(xfs.readdirPromise(storeEntryPath)
+            .then(entries => {
+              return Promise.all(entries.map(async entry => {
+                const p = ppath.join(storeEntryPath, entry);
+                if (entry === Filename.nodeModules) {
+                  const extraneous = await getNodeModulesListing(p);
+                  extraneous.delete(identComponents.join(ppath.sep) as PortablePath);
+                  return cleanNodeModules(p, extraneous);
+                } else {
+                  return xfs.removePromise(p);
+                }
```

Comment:
Just to be sure, this is just to account for "legacy" pnpm installs with the previous store layout?

Reply:
It's to account for any files that shouldn't be there. It has the consequence that it cleans files that were there because of the previous store layout, but it's not specifically for that (so it's not going to be removed once the previous store layout stops being used).